dattri/algorithm/influence_function.py

```diff
@@ -5,99 +5,154 @@
-) -> Tuple[Tensor, List[Tuple[Tensor, ...]]]:
-    """Collate function for LISSA.
+class BaseInnerProductAttributor(BaseAttributor):
+    """The base class for inner product attributor."""
 
-    Args:
-        sampled_input (List[Tensor]): The sampled input from the dataloader.
+    def __init__(
+        self,
+        task: AttributionTask,
+        device: Optional[str] = "cpu",
+        **transformation_kwargs: Dict[str, Any],
+    ) -> None:
+        """Initialize the attributor.
 
-    Returns:
-        Tuple[Tensor, List[Tuple[Tensor, ...]]]: The collated input for the LISSA.
-    """
-    return (
-        sampled_input[0],
-        tuple(sampled_input[i].float() for i in range(1, len(sampled_input))),
-    )
+        Args:
+            task (AttributionTask): The task to be attributed. The task should
+                be an instance of `AttributionTask`.
+            transformation_kwargs (Optional[Dict[str, Any]]): The keyword arguments for
+                the transformation function.
+            device (str): The device to run the attributor.
+        """
+        self.task = task
+        self.transformation_kwargs = transformation_kwargs or {}
+        self.device = device
+        self.index = 0
 
+    def _set_test_data(self, dataloader: torch.utils.data.DataLoader) -> None:
+        """Set test dataloader.
 
-SUPPORTED_IHVP_SOLVER = {
-    "explicit": ihvp_explicit,
-    "cg": ihvp_cg,
-    "arnoldi": ihvp_arnoldi,
-    "lissa": partial(ihvp_lissa, collate_fn=_lissa_collate_fn),
-}
+        Args:
+            dataloader (DataLoader): The dataloader for test samples to be attributed.
+        """
+        # This function may be overrided by the subclass
+        self.test_dataloader = dataloader
 
+    def _set_train_data(self, dataloader: torch.utils.data.DataLoader) -> None:
+        """Set train dataloader to be attributed.
 
-class IFAttributor(BaseAttributor):
-    """Influence function attributor."""
+        Args:
+            dataloader (DataLoader): The dataloader for train samples to be attributed.
+        """
+        # This function may be overrided by the subclass
+        self.train_dataloader = dataloader
 
-    def __init__(
+    def _set_full_train_data(self, dataloader: torch.utils.data.DataLoader) -> None:
+        """Set train dataloader for full training set.
+
+        Args:
+            dataloader (DataLoader): The dataloader for full training samples.
+        """
+        # This function may be overrided by the subclass
+        self.full_train_dataloader = dataloader
+
+    def generate_test_query(
         self,
-        target_func: Callable,
-        params: dict,
-        ihvp_solver: str = "explicit",
-        ihvp_kwargs: Optional[Dict[str, Any]] = None,
-        device: str = "cpu",
-    ) -> None:
-        """Influence function attributor.
+        index: int,
+        data: Tuple[torch.Tensor, ...],
+    ) -> torch.Tensor:
+        """Calculating the query based on the test data.
```

Comment:
quite vague for me of the term: "query". I don't think this is a standard terminology in the context of IF?
Should we borrow some intro from `transformation_on_query`?

Reply:
I guess this one is kind of getting more and more popular in the TDA community.

I will add some intro from `transformation_on_query` here.